vm/actor/src/builtin/power/policy.rs

```diff
@@ -1,107 +1,66 @@
-}
-
-// Penalty to pledge collateral for repeated failure to prove storage.
-pub(super) fn pledge_penalty_for_windowed_post_failure(
-    _pledge: &TokenAmount,
-    _term_type: SectorTermination,
-) -> TokenAmount {
-    // PARAM_FINISH
-    TokenAmount::from(0u8)
-}
-
-/// Penalty to pledge collateral for a consensus fault.
-pub(super) fn pledge_penalty_for_consensus_fault(
-    pledge: TokenAmount,
-    fault_type: ConsensusFaultType,
-) -> TokenAmount {
-    // PARAM_FINISH: always penalise the entire pledge.
-    match fault_type {
-        ConsensusFaultType::DoubleForkMining => pledge,
-        ConsensusFaultType::ParentGrinding => pledge,
-        ConsensusFaultType::TimeOffsetMining => pledge,
-    }
+/// DealWeight and VerifiedDealWeight are spacetime occupied by regular deals and verified deals in a sector.
+/// Sum of DealWeight and VerifiedDealWeight should be less than or equal to total SpaceTime of a sector.
+/// Sectors full of VerifiedDeals will have a SectorQuality of VerifiedDealWeightMultiplier/BaseMultiplier.
+/// Sectors full of Deals will have a SectorQuality of DealWeightMultiplier/BaseMultiplier.
+/// Sectors with neither will have a SectorQuality of BaseMultiplier/BaseMultiplier.
+/// SectorQuality of a sector is a weighted average of multipliers based on their propotions.
+fn sector_quality_from_weight(weight: &SectorStorageWeightDesc) -> SectorQuality {
+    let sector_space_time = weight.sector_size as u64 * weight.duration;
+    let total_deal_space_time = &weight.deal_weight + &weight.verified_deal_weight;
+    assert!(BigInt::from(sector_space_time) < total_deal_space_time);
+
+    let weighted_base_space_time = (sector_space_time - total_deal_space_time) * &*BASE_MULTIPLIER;
+    let weighted_deal_space_time = &weight.deal_weight + &*DEAL_WEIGHT_MULTIPLIER;
+    let weighted_verified_space_time =
+        &weight.verified_deal_weight * &*VERIFIED_DEAL_WEIGHT_MULITPLIER;
+    let weighted_sum_space_time =
+        weighted_base_space_time + (weighted_deal_space_time + weighted_verified_space_time);
+    let scale_up_weighted_sum_space_time =
+        weighted_sum_space_time << ToPrimitive::to_usize(&*SECTOR_QUALITY_PRECISION).unwrap();
+
+    ((scale_up_weighted_sum_space_time / sector_space_time) / &*BASE_MULTIPLIER)
+        .to_biguint()
+        .unwrap()
 }
 
-lazy_static! {
-    static ref INITIAL_SLASHER_SHARE_NUM: BigInt = BigInt::from(1);
-    static ref INITIAL_SLASHER_SHARE_DENOM: BigInt = BigInt::from(1000);
-    static ref SLASHER_SHARE_GROWTH_RATE_NUM: BigInt = BigInt::from(102_813);
-    static ref SLASHER_SHARE_GROWTH_RATE_DENOM: BigInt = BigInt::from(100_000);
+pub fn qa_power_for_weight(weight: &SectorStorageWeightDesc) -> StoragePower {
+    let qual = sector_quality_from_weight(weight);
+    let sector_quality = ToPrimitive::to_usize(&(weight.sector_size as usize * qual)).unwrap();
+    (&*SECTOR_QUALITY_PRECISION >> sector_quality)
+        .to_biguint()
+        .unwrap()
```

Comment:
```suggestion
    let sector_quality = BigUint::from(weight.sector_size as u64) * qual;
    sector_quality >> SECTOR_QUALITY_PRECISION
```
You have the shift backwards, this change is predicated on the change on that const to usize (I checked it works)